best_practices.md

```diff
@@ -0,0 +1,262 @@
+
+## 9. Move constructor should not throw. Mark Move Constructors as `noexcept`
+
+### Practice
+Move constructors should be marked as `noexcept`, otherwise many STL optimizations will not be applied.
+
+### Explanation
+The `noexcept` specifier indicates that a function does not throw exceptions. Marking move constructors as `noexcept` allows the Standard Library to perform various optimizations, such as using move operations in containers more effectively.
+
+### Motivation
+- **Performance**: Enables STL optimizations that rely on `noexcept` guarantees.
+- **Safety**: Clearly indicates that move operations are safe and will not throw exceptions.
+- **Efficiency**: Enhances the performance of STL containers by allowing move operations.
+
+### Example
+https://stackoverflow.com/questions/28627348/why-does-a-throwing-move-constructor-result-in-copying-instead-of-moving-where-a
+
+## 10. Use the Copy-and-Swap Idiom
+
+### Practice
+Use the Copy-and-Swap idiom to avoid duplicating code between different constructors and assignment operators.
+
+### Explanation
+The Copy-and-Swap idiom is a robust and elegant method to implement copy assignment operators. It leverages the copy constructor and the swap method to provide strong exception safety and reduce code duplication.
+
+### Example 
+https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom
+
+
+## 11. Avoid Global Classes, Especially with Mutexes or Locks
+
+### Practice
+Avoid creating global classes, especially those with mutexes or other locks, except for specific debug purposes.
+
+### Explanation
+Global classes can lead to issues with concurrency and resource management, especially when they involve mutexes or other synchronization mechanisms. Such global resources can create hidden dependencies and make the code harder to reason about and maintain.
+
+### Motivation
+- **Concurrency Safety**: Reduces the risk of deadlocks and race conditions.
+- **Maintainability**: Avoids hidden dependencies and makes the code easier to understand and maintain.
+- **Predictability**: Improves the predictability of resource management and execution flow.
+
+## 12. Move Function  Implementations to `.cpp` Files
+
+### Practice
+Move function and method implementations from header files to `.cpp` files. Template heavy functions/methods implementations could be moved to the '_inl.hpp' file which should be included in the end of the header.
+
+### Explanation
+Implementing functions in `.cpp` files rather than in headers reduces compilation dependencies and improves encapsulation. It also leads to faster compile times and reduces the chances of multiple definition errors.
+
+### Motivation
+- **Compilation Time**: Reduces compilation times by minimizing dependencies.
+- **Encapsulation**: Improves encapsulation by hiding implementation details.
+- **Maintainability**: Simplifies code maintenance by separating interface from implementation.
+
+## 13. Avoid `using namespace std;` in Headers
```

Comment:
Maybe "Avoid `using namespace` in headers"?

Reply:
But I will mention that using namespace std is the worst.